### DISCUSSION TITLE
Issues implementing .findIndex()

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#130] [Help Me!] [answered]
Gabby Tobermann 
Searched other tickets: Yes, I am using the solution from ticket: https://github.com/nss-evening-cohort-14/evening-client-side/discussions/65

Issue: (This is for bear watcher assignment) I added a function "attempt" which updates the number of times a bear has attempted to catch a fish. It executes when the "attempted" button is clicked on the card. It was working, but the number of attempt would change on EVERY card, instead a specific one. I added an id on each card, just like Dani's ticket explains. So far I understand how the unique ids work. In my attempt function, I wanted to find the index of the card being targeted. This is where I am confused. I get an error "Uncaught TypeError: Cannot create property 'attempted' on number '-1" 
I know the issue is something about how I am targeting the id inside findIndex().

errors: "Uncaught TypeError: Cannot create property 'attempted' on number '-1"

Google links: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex
I read the documentation again but did not clarify my problem. 
https://ethereum.stackexchange.com/questions/78233/typeerror-cannot-create-property-constant-on-number-123-web3js
When I google the error, I can only find solutions to non-related problems. 
https://stackoverflow.com/questions/13964155/get-javascript-object-from-array-of-objects-by-value-of-property
This one talks about various methods you can use but still doesn't fix my problem. 

What I have tried: replacing findIndex with find() but that just returns undefined.
just writing "const bearIndex = bears.find((bear) => bear.id). This doesn't throw an error, but no matter which "attempt" button I clicked, it will update the first card only. 
console.warn(bearIndex) which returns -1. 

Here is my file: https://github.com/Gabrielle-Tobermann/ASSIGNMENT-Bear-Watcher/blob/timestamp/src/javascripts/Helpers/Data/bears.js

**Comment by Greenfin17** [comment]
Line 32 runs only if e.target.id equals 'attempted', but targetId is assigned the value of e.target.id. Therefore, line 32 only runs if targetId equals 'attempted' and you are getting -1, not found, from bears.find(), since 'attempted' is never one of your bear.id's.

There should be a way to find the bear.id from the card using 'e' since you have added the bear.id to your card.  Try searching the properties of 'e' to find the id. The event 'e' has properties such as parentElement or parentNode.  You can look at your card html to help figure out the relationship between the button element in the card and the card element which has the bear's unique id.

Also, if your button has an id of "attempted" on every card, then it is not unique as an html id should be.

**Reply by drteresavasquez** [reply]
This is great advice. @Gabrielle-Tobermann as John states your target id may not be correct so you need to make sure you're getting the right value for that.

**Comment by drteresavasquez** [comment]
findIndex returns -1 if the item does not exist in the array which means it can't find that object. Like John said, you're comparing a string with the id you've assigned on the bear and that will never be accurate since the id you are looking for is attempted and the bear ids seem to be sequential integers. 

Put the bear ID on the button in your card loop along with the purpose of the button like this: `attempted--${arr[i].id}`

Then do a split on the -- so you can get the id for the bear and if the e.target.id .includes attempted. 

Also remember a query selector only grabs the first item it finds so all your ids need to be unique and doing this will make it that way.

**Comment by Gabrielle-Tobermann** [comment]
Thanks @Greenfin17  and @drteresavasquez for your help. So I changed the ID to `attempted--${arr[i].id` on the button and applied the split method to get the id. It is working properly, when I console.warn() the variable holding the id, it is logging the correct id. But I am still having issues with the findIndex method. It is still returning -1. Here is what my code looks like now: 
const attempt = (e) => {
  const targetId = e.target.id;
  if (targetId.includes('attempted')) {
    const bearId = targetId.split('--').slice(-1);
    const bearIdNum = bearId[0];
    const bearIndex = bears.findIndex((bear) => bear.id === bearIdNum);
    bearIndex.attempted += 1;
    bearIndex.timeAttempt = new Date();
  }
  createCards(bears);
};

I thought it might be returning -1 because bearId was an array, so I added bearIdNum variable so it would be a number. I also tried to use double operator instead of '===' but this is not esLint friendly.

**Reply by drteresavasquez** [reply]
Try this and log it to make sure you get a number back. 

`Number(targetId.split('--')[1]);`

**Reply by Gabrielle-Tobermann** [reply]
It's working! Thank you! I will add a comment with the solution.

**Reply by drteresavasquez** [reply]
Yassss!!!!! 💪🏽

**Comment by Gabrielle-Tobermann** [comment, marked as answer]
To make the id unique, I changed it to `attempted--${arr[i].id}` . Then to extract the id, I created a variable bearId that would store only the id part, like this: `const bearId = Number(targetId.split('--')[1]);`. It also converts the value to a number. 
Then, to find the index of the card with the correct id: `const bearIndex = bears.findIndex((bear) => bear.id === bearId);`
To change the value of the attempted key:  `bears[bearIndex].attempted += 1`;
